Source/Core/CesiumTerrainProvider.js

```diff
@@ -840,9 +987,124 @@ define([
+        var divisor = 1 << (level - parentLevel);
+        var parentX = (x / divisor) | 0;
+        var parentY = (y / divisor) | 0;
+
+        return {
+            level: parentLevel,
+            x: parentX,
+            y: parentY
+        };
+    }
+
+    function checkAncestorsLayers(provider, x, y, level, index) {
+        var layers = provider._layers;
+        var layer = layers[index];
+        var promise;
+        if (defined(layer.availabilityLevels)) {
+            // We only need to check this layer if there wasn't
+            //  an available list and it has the metadata extension
+            var parent = getAvailabilityTile(layer, x, y, level);
+            promise = checkParentTiles(provider, parent.x, parent.y, parent.level, layer);
+        }
+
+        // Nothing to load, so this tile isn't available in this layer
+        if (!defined(promise)) {
+            // This layer doesn't have it so we can move on to check the next layer
+            if (++index === layers.length)
+            {
+                // No more layers left
+                return;
+            }
+            return checkAncestorsLayers(provider, x, y, level, index);
+        }
+
+        return promise
+            .then(function() {
+                if (layer.availability.isTileAvailable(level, x, y)) {
+                    // We now know we have this tile, so we can stop here
+                    return;
+                }
+
+                // This layer doesn't have it so we can move on to check the next layer
+                if (++index === layers.length)
+                {
+                    // No more layers left
+                    return;
+                }
+                return checkAncestorsLayers(provider, x, y, level, index);
+            });
+    }
+
+    function checkParentTiles(provider, x, y, level, layer) {
+        var isLoaded = layer.availabilityTilesLoaded.isTileAvailable(level, x, y);
+        if (isLoaded) {
+            // Tile is loaded, so just return
+            return;
+        }
+
+        var promise = when.resolve();
+        if (level !== 0) {
+            var parent = getAvailabilityTile(layer, x, y, level);
+            promise = checkParentTiles(provider, parent.x, parent.y, parent.level, layer);
+        }
+
+        // If all parent tiles are already loaded, then this tile isn't available
+        if (!defined(promise)) {
+            return;
+        }
+
+        return promise
+            .then(function() {
+                if (!layer.availability.isTileAvailable(level, x, y)) {
+                    // All parents are loaded, so if this tile isn't available don't try to load it.
+                    return;
+                }
+
+                // Check the cache
+                var key = level + '-' + x + '-' + y;
+                var promise = layer.availabilityPromiseCache[key];
+                if (defined(promise)) {
+                    return promise;
```

Comment:
It's good there's a cache of promises for tiles loaded for availability purposes, but I think it's employed a bit late. If I'm reading this right, it's still going to create a hella-long promise chain for each position given to `sampleTerrainMostDetailed`. Most of those promise chains won't lead to an actual load because of the cache, but 
when.js promises are still pretty expensive even without a load.

Reply:
IMO, the `sampleTerrainMostDetailed` algorithm should be something more like this:

1. Find the _known available_ tiles that contain our positions.
2. Load them.
3. Check if we've now learned that there are new, relevant available tiles for our terrain sampling.
4. If so, repeat starting with step 1.

I think that makes more sense than assuming that the max level tiles are available, trying to load those, and working up the tree if that fails. It generates less 404s, which take over a second from Ion currently and return no useful information whatsoever. Also, to a first approximation, max level tiles _don't_ exist. If they did (on average), the terrain dataset would be petabytes or something. The availability of the most detailed terrain is quite spotty, assuming otherwise is a bad assumption.

The algorithm above is basically how the rendering should work too, at least in the fill-tiles branch. Missing availability should stop refinement until availability is known. I think creating promises in the course of checking availability is going to cause way too much overhead and we should avoid it. Promises to load tiles are fine, of course.

By the way, how many levels are of availability are you putting in each tile? I reckon with this new format it can be quite high, like 10+, so it won't substantially impact load time to go top-down when loading availability. Am I wrong?

Reply:
This sounds good (and way less complicated to implement).

> By the way, how many levels are of availability are you putting in each tile? I reckon with this new format it can be quite high, like 10+, so it won't substantially impact load time to go top-down when loading availability. Am I wrong?

Right now I'm using 10 levels. I think that will be a good balance of keeping tiles relatively small but maximizing how much availability we can see.